File: be-core/src/main/java/com/dentalCare/be_core/services/OdontogramService.java
package com.dentalCare.be_core.services;

import com.dentalCare.be_core.dtos.request.odontogram.OdontogramRequestDto;
import com.dentalCare.be_core.dtos.response.odontogram.OdontogramResponseDto;

import java.util.List;

public interface OdontogramService {
    
    /**
     * Crear un nuevo odontograma para un paciente
     * @param dentistId ID del dentista que crea el odontograma
     * @param requestDto Datos del odontograma
     * @return Odontograma creado
     */
    OdontogramResponseDto createOdontogram(Long dentistId, OdontogramRequestDto requestDto);
    
    /**
     * Obtener todos los odontogramas de un paciente
     * @param dentistId ID del dentista
     * @param patientId ID del paciente
     * @return Lista de odontogramas del paciente
     */
    List<OdontogramResponseDto> getOdontogramsByPatient(Long dentistId, Long patientId);
    
    /**
     * Obtener un odontograma específico
     * @param odontogramId ID del odontograma
     * @param dentistId ID del dentista
     * @return Odontograma encontrado
     */
    OdontogramResponseDto getOdontogramById(Long odontogramId, Long dentistId);
    
    /**
     * Actualizar un odontograma existente
     * @param odontogramId ID del odontograma
     * @param dentistId ID del dentista
     * @param requestDto Nuevos datos del odontograma
     * @return Odontograma actualizado
     */
    OdontogramResponseDto updateOdontogram(Long odontogramId, Long dentistId, OdontogramRequestDto requestDto);
    
    /**
     * Eliminar (lógicamente) un odontograma
     * @param odontogramId ID del odontograma
     * @param dentistId ID del dentista
     */
    void deleteOdontogram(Long odontogramId, Long dentistId);
    
    /**
     * Contar odontogramas de un paciente
     * @param patientId ID del paciente
     * @return Cantidad de odontogramas activos
     */
    Long countOdontogramsByPatient(Long patientId);
}

File: dental-portal/src/app/features/dentists/services/odontogram.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiConfig } from '../../../core/config/api.config';

export interface ToothData {
  number: number;
  statuses: string[];
}

export interface OdontogramRequestDto {
  patientId: number;
  dentitionType: 'adult' | 'child';
  teethData: string; // JSON string
}

export interface OdontogramResponseDto {
  id: number;
  patientId: number;
  dentitionType: 'adult' | 'child';
  teethData: string; // JSON string
  createdDatetime: string;
  isActive: boolean;
}

@Injectable({
  providedIn: 'root'
})
export class OdontogramService {
  private http = inject(HttpClient);
  private apiConfig = inject(ApiConfig);
  private apiUrl = this.apiConfig.coreDentistUrl;

  /**
   * Crear un nuevo odontograma
   */
  createOdontogram(dentistId: number, patientId: number, requestDto: OdontogramRequestDto): Observable<OdontogramResponseDto> {
    return this.http.post<OdontogramResponseDto>(
      `${this.apiUrl}/${dentistId}/patients/${patientId}/odontogram`,
      requestDto
    );
  }

  /**
   * Obtener todos los odontogramas de un paciente
   */
  getOdontogramsByPatient(dentistId: number, patientId: number): Observable<OdontogramResponseDto[]> {
    return this.http.get<OdontogramResponseDto[]>(
      `${this.apiUrl}/${dentistId}/patients/${patientId}/odontogram`
    );
  }

  /**
   * Obtener un odontograma específico por ID
   */
  getOdontogramById(dentistId: number, odontogramId: number): Observable<OdontogramResponseDto> {
    return this.http.get<OdontogramResponseDto>(
      `${this.apiUrl}/${dentistId}/odontogram/${odontogramId}`
    );
  }

  /**
   * Actualizar un odontograma existente
   */
  updateOdontogram(dentistId: number, odontogramId: number, requestDto: OdontogramRequestDto): Observable<OdontogramResponseDto> {
    return this.http.put<OdontogramResponseDto>(
      `${this.apiUrl}/${dentistId}/odontogram/${odontogramId}`,
      requestDto
    );
  }

  /**
   * Eliminar (lógicamente) un odontograma
   */
  deleteOdontogram(dentistId: number, odontogramId: number): Observable<void> {
    return this.http.delete<void>(
      `${this.apiUrl}/${dentistId}/odontogram/${odontogramId}`
    );
  }

  /**
   * Contar odontogramas de un paciente
   */
  countOdontogramsByPatient(patientId: number): Observable<number> {
    return this.http.get<number>(
      `${this.apiUrl}/patients/${patientId}/odontogram/count`
    );
  }
}
